ls1intum/Artemis
# `Discussion`: Add sorting by date to course and lecture discussion tabs

### Checklist
#### General


#### Client

### Motivation and Context


Follow up PR for #5167. Sorting posts by creation date will enhance the user experience especially when many posts are created for a specific lecture or exercise.


### Description

We further extend the Post sorting mechanism of the exercise & lecture discussion tabs so users are able to sort Posts by ASC/DESC creation date respectfully. The sorting happens as before when the page is initially loaded. Then the user may click the 'arrow' icon found on the right-hand side of the Post filters to switch between sort asc by date, sort desc by date, and finally default sorting.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 course with postings enabled
- 1 exercise or lecture

1. Log in to Artemis
2. Navigate to an exercise or a lecture
3. Verify that the following criteria are applied while sorting posts;

- 1. criterion: displayPriority is PINNED -> pinned posts come first
- 2. criterion: displayPriority is ARCHIVED  -> archived posts come last
-- in between pinned and archived posts --
- 3. criterion (optional): creationDate - if activated by user through the sort arrow -> most recent comes at the end (chronologically from top to bottom)
- 4. criterion: if 3'rd criterion was not activated by the user, vote-emoji count -> posts with more vote-emoji counts comes first

!!! UPDATED !!!

- 5. criterion: ~if 3'rd criterion was not activated by the user,~ most recent posts comes at the end (chronologically from top to bottom)

!!! UPDATED !!!

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
course-discussion.component.ts | 80.24% | 95.18% |
discussion-section.component.ts | 83.78% | 85.41% |
course-discussion.directive.ts | 91.66% | 100% |


### Screencast


https://user-images.githubusercontent.com/67344158/171614263-e49b5b68-76b9-4c94-a170-612bb5332c0f.mp4
